sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/coder/KryoCoder.java

```diff
@@ -18,87 +18,107 @@
-    final byte[] bytes;
-
-    try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        final Output output = new Output(baos)) {
-      kryo.get().writeClassAndObject(output, t);
-      output.flush();
-      bytes = baos.toByteArray();
-    }
-    final DataOutputStream dos = new DataOutputStream(out);
-    dos.writeInt(bytes.length);
-    dos.write(bytes);
+  private KryoCoder(IdentifiedRegistrar registrarWithId) {
+    this.registrarWithId = registrarWithId;
+  }
+
+  /**
+   * Creates a {@link KryoCoder} instance which will use {@link Kryo} without class registration.
+   * That degrades performance. Use {@link #of(IdentifiedRegistrar)} whenever possible.
+   */
+  public KryoCoder() {
+    this(KryoFactory.NO_OP_REGISTRAR);
+  }
+
+  /**
+   * Creates a {@link KryoCoder} instance which will use {@link Kryo} with classes registered by
+   * {@code registrar}.
+   */
+  public static <T> KryoCoder<T> of(IdentifiedRegistrar registrarWithId) {
+    return new KryoCoder<>(registrarWithId);
   }
 
   @Override
-  @SuppressWarnings("unchecked")
-  public T decode(InputStream is) throws IOException {
-
-    final DataInputStream dis = new DataInputStream(is);
-    final int size = dis.readInt();
-    final byte[] buffer = new byte[size];
-    if (size != dis.read(buffer, 0, size)) {
-      throw new IllegalStateException("This should never happen.");
-    }
-    try (final InputStream in = new ByteArrayInputStream(buffer);
-        final Input input = new Input(in)) {
+  public void encode(T value, OutputStream outStream) throws IOException {
 
-      return (T) kryo.get().readClassAndObject(input);
+    Output output = KryoFactory.getKryoOutput();
+    output.clear();
+    Kryo kryo = KryoFactory.getOrCreateKryo(registrarWithId);
+    try {
+      kryo.writeClassAndObject(output, value);
+    } catch (IllegalArgumentException e) {
+      throw new CoderException(
+          String.format(
+              "Cannot encode given object of type '%s'. Forgotten kryo registration is possible explanation.",
+              (value == null) ? null : value.getClass().getSimpleName()),
+          e);
     }
+
+    DataOutputStream dos = new DataOutputStream(outStream);
+    // write length of encoded object first to get ability to limit read in decode() to one object
+    dos.writeInt(output.position());
+    outStream.write(output.getBuffer(), 0, output.position());
   }
 
   @Override
-  public void verifyDeterministic() throws NonDeterministicException {
-    // nop
-  }
+  public T decode(InputStream inStream) throws IOException {
+
+    // read length of objects in bytes first
+    DataInputStream dis = new DataInputStream(inStream);
+    int lengthOfDecodedObject = dis.readInt();
+
+    Input input = KryoFactory.getKryoInput();
+
+    // limit the input stream to not allow kryo to read beyond the current object
+    BoundedInputStream limitedStream = new BoundedInputStream(inStream, lengthOfDecodedObject);
```

Comment:
👍 I'm just litle bit concerned about apache commons dependency? Where did it come from?

Reply:
Good point. It will be removed in [BEAM-5179](https://issues.apache.org/jira/browse/BEAM-5179). When Kryo's chunked in/out will replace our own chunking.